src/plugins/intel_cpu/src/nodes/kernels/registers_pool.hpp

```diff
@@ -0,0 +1,184 @@
+// Copyright (C) 2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include "cpu/x64/jit_generator.hpp"
+#include <dnnl_types.h>
+
+namespace ov {
+namespace intel_cpu {
+
+using namespace dnnl::impl::cpu;
+
+/*
+ * The RegistersPool is the base class for the RegPool template:
+ *      template <x64::cpu_isa_t isa>
+ *      class RegPool : public RegistersPool;
+ *
+ * The registers pool must be created by instantiating the RegPool template, like the next:
+ *      RegPool<isa> regPool { // the list of the registers to be excluded from pool
+ *          Reg64(Operand::RAX), Reg64(Operand::RCX), Reg64(Operand::RDX), Reg64(Operand::RBX),
+ *          Reg64(Operand::RSP), Reg64(Operand::RBP), Reg64(Operand::RSI), Reg64(Operand::RDI)
+ *      };
+ */
+class RegistersPool {
+public:
+    /*
+     * The scoped wrapper for the Xbyak registers.
+     * By creating it you are getting the register from the pool RegistersPool.
+     * It could be created by using constructor with RegistersPool as an argument, like the next:
+     *      const RegistersPool::Reg<Xbyak::Reg64> reg {regPool};
+     * The destructor will return the register to the pool.
+     */
+    template<class TReg>
+    class Reg {
+        friend class RegistersPool;
+    public:
+        Reg(RegistersPool& regPool);
+        ~Reg() { if (onDestruct) { onDestruct(); } }
+        Reg& operator=(Reg&& other)  noexcept {
+            reg = other.reg;
+            onDestruct = std::move(other.onDestruct);
+            return *this;
+        }
+        Reg(Reg&& other)  noexcept : reg(other.reg), onDestruct(std::move(other.onDestruct)) {}
+        operator TReg&() { assert(isValid()); return reg; }
+        operator const TReg&() const { assert(isValid()); return reg; }
+        operator Xbyak::RegExp() const { assert(isValid()); return reg; }
+        int getIdx() const { assert(isValid()); return reg.getIdx(); }
+        friend Xbyak::RegExp operator+(const Reg& lhs, const Xbyak::RegExp& rhs) {
+            assert(lhs.isValid());
+            return lhs.operator Xbyak::RegExp() + rhs;
+        }
+        bool isValid() const { return onDestruct.operator bool(); }
+
+    private:
+        Reg(const TReg& reg, const std::function<void()>& onDestruct)
+                : reg(reg), onDestruct(onDestruct) {}
+
+    private:
+        TReg reg;
+        std::function<void()> onDestruct;
+    };
+
+    virtual ~RegistersPool() = default;
+
+protected:
+    struct PhysicalSet {
```

Comment:
Please, make it as class and hide member fields as private

Reply:
Done